[FLINK-25575] Add Sink V2 operators and translation

fapaul opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This PR allows executing Sink V2 and migrates the internal operator representation of Sink V1 to also use the Sink V2 operators. So basically every Sink V1 now runs internally on the Sink V2 infrastructure. Only the GlobalCommitter is currently still missing but it is tracked as a followup in https://issues.apache.org/jira/browse/FLINK-25726.

## Brief change log

- d45e3a669eece2fc287d454ed4d69a5a34c9dff9 This adds the internal structure send and book keep the Committable Messages it ensures that the operators working with CommittableMessages can ensure exactly-once processing
- 029d72ebef1e5df84313ea9c23e86ee0ccaecda7 small change to expose the held transformations of the stream environment. It is mainly used later by the SinkTransformationTranslator to ingest newly added transformations to the StreamGraphGenerator
- 90556071ebceb2a01768bae0db671d2a808f8e23 Batch/Blocking exchanges are now changed by the framework accordingly in Streaming Mode. They only have an effect in Batch Mode.
- fa0eaa1541ce6d05fe8460c060c45cb0f6bb196d Adds a simulation adapter to convert Sink V1 to Sink V2.
- f799eeea34b4d3c5691cc4b77f5be103ef1c065c Implements the operator logic and adapts the SinkTransformationTranslation. Sorry for that commit it is very huge.

## Verifying this change

Added multiple new tests for the new operators and for the adapter simulation. In general, all Sink IT cases also verify that the Sink V2 works.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

alpreu commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    /**
+     * Returns whether all {@link CheckpointCommittables} currently hold by the collector are either
```
```suggestion
     * Returns whether all {@link CheckpointCommittables} currently held by the collector are either
```

fapaul commented:
@flinkbot run azure

JingGe commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java:
```diff
@@ -0,0 +1,235 @@
+     *
+     * @param message either {@link CommittableSummary} or {@link CommittableWithLineage}
+     */
+    public void addMessage(CommittableMessage<CommT> message) {
```
Nit

fapaul replied:
I'd prefer to keep it short.

JingGe replied:
sure, it is not a big deal. The long name will improve the code readability at the place where the method is called, since code like `addMessage(message)` does not contain enough info. But once the developer get familiar with the code, it is not an issue.

JingGe reviewed (commented):
Thanks for your contribution! Since the PR is huge, I will do the review in stages following the commit sequence. 

Left some comments for the commit d45e3a669eece2fc287d454ed4d69a5a34c9dff9

fapaul commented on flink-architecture-tests/violations/5b9eed8a-5fb6-4373-98ac-3be2a71941b8:
```diff
@@ -93,7 +93,6 @@ org.apache.flink.connector.file.src.util.Pool.recycler(): Returned leaf type org
 org.apache.flink.connector.file.src.util.Utils.forEachRemaining(org.apache.flink.connector.file.src.reader.BulkFormat$Reader, java.util.function.Consumer): Argument leaf type org.apache.flink.connector.file.src.reader.BulkFormat$Reader does not satisfy: reside outside of package 'org.apache.flink..' or annotated with @Public or annotated with @PublicEvolving or annotated with @Deprecated
 org.apache.flink.connector.jdbc.JdbcExactlyOnceOptions.builder(): Returned leaf type org.apache.flink.connector.jdbc.JdbcExactlyOnceOptions$JDBCExactlyOnceOptionsBuilder does not satisfy: reside outside of package 'org.apache.flink..' or annotated with @Public or annotated with @PublicEvolving or annotated with @Deprecated
 org.apache.flink.connector.jdbc.JdbcExecutionOptions.builder(): Returned leaf type org.apache.flink.connector.jdbc.JdbcExecutionOptions$Builder does not satisfy: reside outside of package 'org.apache.flink..' or annotated with @Public or annotated with @PublicEvolving or annotated with @Deprecated
-org.apache.flink.connector.jdbc.JdbcSink.exactlyOnceSink(java.lang.String, org.apache.flink.connector.jdbc.JdbcStatementBuilder, org.apache.flink.connector.jdbc.JdbcExecutionOptions, org.apache.flink.connector.jdbc.JdbcExactlyOnceOptions, org.apache.flink.util.function.SerializableSupplier): Argument leaf type org.apache.flink.util.function.SerializableSupplier does not satisfy: reside outside of ...
```
I guess this change should have belonged to this ticket https://issues.apache.org/jira/browse/FLINK-25570 I am not sure why the tests did not fail before. I can make this change on a separate commit but I would keep it on this PR.

For the other archunit changes afaict they are related to changes of this PR.

JingGe replied:
I have the same issue like you have. It seems there is another PR will take the responsibility for it. FYI: https://github.com/apache/flink/pull/18333#discussion_r791587285

fapaul replied:
Thanks for the hint, I'll remove my change as I think the other PR is merged before this one.

Airblader replied:
It's merged now. Sorry for that!

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittables.java:
```diff
@@ -0,0 +1,36 @@
+ * @param <CommT> type of the committable
+ */
+@Internal
+public interface CheckpointCommittables<CommT> extends Committables<CommT> {
```
The idea behind the two interfaces is to have one abstraction for batch execution `CommittableManager` that does not have a checkpoint Id because there a no checkpoint and `CheckpointCommittableManager` for streaming use-cases where checkpoints are supported.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/SubtaskCommittables.java:
```diff
@@ -0,0 +1,168 @@
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+class SubtaskCommittables<CommT> {
```
Formerly it is not required to have a doc string for `package-private` classes.

JingGe replied:
> Formerly it is not required to have a doc string for `package-private` classes.

I feel that SubtaskCommittable is one of the most important implementation classes. TBH, `package-privat` should not be the criteria for adding java doc or not. Java doc is required if the code can not explain itself or it will take too much time to read through the code, even it is a private method.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    private CheckpointCommittablesImpl<CommT> getCheckpointCommittables(
+            CommittableMessage<CommT> committable) {
```
If I change the signature of the method I need to expose the `EOI` (end of input) checkpoint number to the outside. I think it is best to keep it contained in this class.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittablesImpl.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    CheckpointCommittablesImpl(
+            Map<Integer, SubtaskCommittables<CommT>> subtasksCommittables, long checkpointId) {
```
I could actually happen that the `subtaskCommittables` have lower checkpoint Ids because in Flink it can happen that a checkpoint subsumes a previous one.

JingGe replied:
good point, I would suggest to add this info as java doc.

fapaul replied:
I added a java doc

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittablesImpl.java:
```diff
@@ -0,0 +1,141 @@
+        return subtasksCommittables.values();
+    }
+
+    void addSummary(CommittableSummary<CommT> summary) {
```
I think from a caller perspective it adds the summary and upsert would show some implementation detail.

JingGe replied:
There is update logic in the add method, i.e. hidden magic code.

fapaul replied:
I renamed it to `upserSummary`

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommitRequestImpl.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public void signalFailedWithKnownReason(Throwable t) {
+        state = CommitRequestState.FAILED;
+        // add metric later
```
I opened a ticket to implement the metrics https://issues.apache.org/jira/browse/FLINK-25857

fapaul commented:
thanks for your first review @JingGe I have addressed all your comments.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/Committables.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * Commits all due committables.
+     *
+     * @param fullyReceived only commit committables if all committables of this checkpoint for a
```
Do you have a suggestion here? I am having a hard time to find a better phrasing.

JingGe replied:
I am not sure Whether I understood your idea correctly, for example: 

@param fullyReceived gives the information whether all committables for the checkpoint are received, please refer to {@link CommitterOperator} for more information. Implementation class could use it with its own logic to determine the pending comitRequests that need to be committed, for example to check further if subtasksCommittableManager has received all commitables.

fapaul replied:
`fullyReceived` is more like a signal that committables coming from a certain subtask for a checkpointId should be committed although not all committables from this subtask/checkpointId have been received.

JingGe commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:
```diff
@@ -17,95 +17,173 @@
+    private void commitAndEmitCheckpoints() throws IOException, InterruptedException {
+        for (CheckpointCommittables<CommT> checkpoint :
+                committableCollector.getCheckpointCommittablesUpTo(lastCompletedCheckpointId)) {
+            // wait for all committables of the current checkpoint before submission
```
the comment is a little bit confused. What does a boolean check mean "wait"?

fapaul replied:
I guess this one is related to the `fullyReceived` discussion. The idea is that we do not want to send out committables of the latest checkpoint if we haven't received all of them. Does this make sense?

JingGe reviewed (commented):
This PR is huge. Thanks again for your effort! 

enclosed please find some more comments for the rest commits

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/PartitionTransformationTranslator.java:
```diff
@@ -70,13 +73,15 @@
 
+        StreamExchangeMode exchangeMode = transformation.getExchangeMode();
+        if (!supportsBatchExchange && exchangeMode == StreamExchangeMode.BATCH) {
+            exchangeMode = StreamExchangeMode.UNDEFINED;
```
Basically in streaming mode, there is no handling for `BATCH` exchange mode they are only relevant in batch execution mode. Therefore it is safe to reset the edges in streaming mode. I'll add a comment

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java:
```diff
@@ -0,0 +1,447 @@
+        }
+    }
+
+    private class StatefulTwoPhaseCommittingSinkAdapter extends StatefulSinkAdapter
```
I am not sure I understand your idea. The problem is that previously we only had one interface `Sink<InputT, CommT, WriterStateT, GlobalCommT>`  that is now represented by multiple mixins. So I had to split the different use cases into different classes implementing the new interfaces.

JingGe replied:
ISP(Interface Segregation Principle) is controlled by the interface not the implementation classes. It is fine to split the big old Sink interface based on ISP. Using one class to implement them all will not violate it. Because what the caller will see are fine-grained interfaces not the implementation class(es), especially when all of these classes are private inner classes and most logic is to wrap each other.

fapaul replied:
AFAICT I need multiple classes or it increases the complexity to implement the given use cases with one class i.e. converting a previously only SinkWriter -> GlobalCommitter topologies now becomes SinkWriter -> ForwardCommitter -> GlobalCommitter. I think the conditional logic is then distributed across all creation methods (i.e. SinkWriter, Commiter, GlobalCommiter) which I find harder to understand instead of doing the checks once and returning different implementations.

gaoyunhaii reviewed (commented):
Thanks @fapaul for drafting the PR! I first left some comments for the translation part. I'll continue to have a look at the other part.

I have one major concern regarding currently we might add the expanded transformations to the original environment. This seems to might cause problems if users call execute multiple times with the same env like

```
DataStream a = ...
DataStream b = ...
b.addSink(new xx());

// The following would first clear all the current
// transformations, but when translating the sink,
// it would add the expanded transformations here.
env.execute()

b.map(...);
// Then here it would execute a graph also with 
// a -> some sink transformations
env.execute()
```
Is it possible we use a separate `env` for the sink translation ? We could wrap the input transformation with a DataStream in the new environment, and at last we get all the transformations and transform them. We could also disable some methods inside adding topology, like switch runtime mode or configure slot resources.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStreamSink.java:
```diff
@@ -40,32 +45,46 @@
     /** Returns the transformation that contains the actual sink operator of this sink. */
     @Internal
-    public Transformation<T> getTransformation() {
+    public Transformation<?> getTransformation() {
```
I reverted that change it feels unncessary.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStreamSink.java:
```diff
@@ -40,32 +45,46 @@
+    }
+
+    static <T> DataStreamSink<T> forSinkFunction(
+            DataStream<T> inputStream, SinkFunction<T> sinkFunction) {
```
This is more like a cleanup that the `DataStreamSink` is the central entry point to add the sink transformation to the datastream. I suspect the `SinkFunction` will be in Flink for a longer time still because it requires Flink 2.0 to remove it.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStreamSink.java:
```diff
@@ -131,6 +150,10 @@ protected DataStreamSink(DataStream<T> inputStream, Sink<T, ?, ?, ?> sink) {
      */
     @PublicEvolving
     public DataStreamSink<T> setUidHash(String uidHash) {
+        if (!(transformation instanceof LegacySinkTransformation)) {
```
No, this exception immediately occurs when the main method is executed before the transformations are translated. It fails the whole job submission.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java:
```diff
@@ -924,5 +924,10 @@ public long getDefaultBufferTimeout() {
         }
+
+        @Override
+        public Collection<Integer> transform(Transformation<?> transformation) {
```
That is definitely not the part I am most proud of :) Unfortunately, I need a way in a `SinkTransformationTranslator` to enqueue new transformations for the translation to support the dynamic expansion in case a sink uses a custom topology. 

The only thing exposed is the `Context` in `translateForStreaming` and `translateForBatch`.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterStateHandler.java:
```diff
@@ -17,34 +17,27 @@
+    void snapshotState(long checkpointId) throws Exception;
+
+    /** Creates a writer, potentially using state from {@link StateInitializationContext}. */
+    SinkWriter<InputT> createWriter(InitContext initContext, StateInitializationContext context)
```
Can you elaborate on why the name is not descriptive? It creates a sink writer instance. I cannot easily change the class name because it is part of the public API and overall this interface is solely intended for internal usage.

JingGe replied:
The class name only tells that it is responsible for handling the state.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/WriterOperator.java:
```diff
@@ -56,105 +59,64 @@
  */
-class SinkOperator<InputT, CommT, WriterStateT> extends AbstractStreamOperator<byte[]>
-        implements OneInputStreamOperator<InputT, byte[]>, BoundedOneInput {
+class WriterOperator<InputT, CommT> extends AbstractStreamOperator<CommittableMessage<CommT>>
```
Usually, I am not a big fan to encode duplicate information in the java module path and class name. In this case, all the classes are already in the `sink` module and therefore belong to the `sink`. On the other, we also only call it `CommitterOperator` and not `SinkCommitterOperator`. 

The information you are looking for should be part of the public-facing interfaces (`SinkWriter`, `Committer`) and I do not think it makes much sense to repeat them internally.

JingGe replied:
From the code readability perspective, first of all, the package info is "hidden" in the import corner; second, it is not consistent, `Committer` -> `CommitterOperator`, but `SinkWriter` -> `WriterOperator`. Comparing to "Committer", the word "Writer" is more common and abstracted, could be used in many different scenarios. But, your point is also fair enough.

fapaul replied:
I renamed to `SinkWriterOperator` to be consistent.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkTransformation.java:
```diff
@@ -71,47 +68,23 @@ public void setChainingStrategy(ChainingStrategy strategy) {
-        result.add(this);
-        result.addAll(input.getTransitivePredecessors());
-        return result;
+        return Lists.newArrayList();
```
Good catch. I think it is safe to revert this.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java:
```diff
@@ -7,7 +7,7 @@
  * with the License.  You may obtain a copy of the License at
  *
- * http://www.apache.org/licenses/LICENSE-2.0
+ *     http://www.apache.org/licenses/LICENSE-2.0
```
This was changed automatically by my IDE because I touched the file. In general, it seems our license includes this indentation [1], and probably the next one touching this file runs into the same issue. Do you think I should make a separate commit for this change? 

[1] https://nightlies.apache.org/flink/flink-docs-master/docs/flinkdev/ide_setup/#copyright-profile

gaoyunhaii replied:
Very thanks for the explanation~ It seems we have many classes with the inconsistency copyright, thus perhaps we fix it in a separate PR ?

fapaul replied:
Unfortunately every time I save it in the IDE (Intellij) it is automatically updated 🙈

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java:
```diff
@@ -19,250 +19,237 @@
+                if (ssg.isPresent() && !subTransformation.getSlotSharingGroup().isPresent()) {
+                    subTransformation.setSlotSharingGroup(ssg.get());
+                }
+                subTransformation.setParallelism(transformation.getParallelism());
```
Is there a good way to determine that the user explicitly set the parallelism? So far I assumed that if the user sets the parallelism of the `SinkTransformation` it applies to all transformations that are expanded.

I also do not see a good a way to distinguish between "user set" parallelism because afaik when a `Transformation` is created you always have to pass a parallelism.

In end, we probably have to define our own interpretation of what setting the parallelism at the `SinkTransformation` implies.

gaoyunhaii replied:
It seems we indeed do not have a good method to check if parallelism is set. Logically we should have the priority with `customized > sink transformations' settings > env default`. We might have to support it since for compaction we need to set the compact coordinator to parallelism 1...

One possible solution is still to use a separate environment with the default parallelism set to sink transformation's settings...

gaoyunhaii replied:
Hi Fabian~I'm a bit afraid there might be bad cases for this method since users might
```
env.setParallelism(1);
DataStream x = ...
x.addSink(new FinkSink()).setParallism(2)
```

while inside the precommit topology, operators like CompactorCoordiantor is set to 1 specially.

In this case, the final parallelism for the compactor coordinator would still be 2 and break the priority?

fapaul replied:
Hopefully, I found a final solution now I had to change how the parallelism is passed to the transformations 06eb49a. Now defaulting to the env parallelism is the responsibility of the translators and by default, the parallelism is always `ExecutionConfig#DEFAULT_PARALLELISM`. This allows to do:

```java
                if (subTransformation.getParallelism() == ExecutionConfig.PARALLELISM_DEFAULT) {
                    if (transformation.getParallelism() == ExecutionConfig.PARALLELISM_DEFAULT) {
                        subTransformation.setParallelism(
                                context.getStreamGraph().getExecutionConfig().getParallelism());
                    } else {
                        subTransformation.setParallelism(transformation.getParallelism());
                    }
                }

```

fapaul commented:
@gaoyunhaii thanks for your review. 

> I have one major concern regarding currently we might add the expanded transformations to the original environment. This seems to might cause problems if users call execute multiple times with the same env like
> 
> ```
> DataStream a = ...
> DataStream b = ...
> b.addSink(new xx());
> 
> // The following would first clear all the current
> // transformations, but when translating the sink,
> // it would add the expanded transformations here.
> env.execute()
> 
> b.map(...);
> // Then here it would execute a graph also with 
> // a -> some sink transformations
> env.execute()
> ```
> 
> Is it possible we use a separate `env` for the sink translation ? We could wrap the input transformation with a DataStream in the new environment, and at last we get all the transformations and transform them. We could also disable some methods inside adding topology, like switch runtime mode or configure slot resources.

Maybe I overlooked something but I thought after every `StreamExecutionEnvironment#execute` the transformations are cleared. [1][2][3]

Regarding your failover region comment https://github.com/apache/flink/pull/18428#discussion_r794246463 (for some reason I cannot answer your comment in the UI...) I assumed that the Batch exchange will never have an effect in streaming mode and therefore we could handle them equally. If that is not the case I also tend to do your suggestion.

[1] https://github.com/apache/flink/blob/1704b1336bcbcdd42405f5e6e3e09c0176925210/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java#L2137
[2] https://github.com/apache/flink/blob/1704b1336bcbcdd42405f5e6e3e09c0176925210/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java#L2120
[3] https://github.com/apache/flink/blob/1704b1336bcbcdd42405f5e6e3e09c0176925210/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java#L1950

gaoyunhaii commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/SubtaskCommittableManager.java:
```diff
@@ -0,0 +1,208 @@
+
+/** Manages the committables coming from one subtask. */
+class SubtaskCommittableManager<CommT> {
+    private CommittableSummary<CommT> summary;
```
Could we directly use a `int numCommittables` or `int numTotal` to bookkeep the total number of the committables? 

The only usage for other fields is the `drainCommitted` method below. I'm a bit concern on the right `subtaskId` to fill in here: if we want to keep the same protocol for the post-committer topology, the subtask id here should be the id of the current commit operator subtask? (Assume users might create operators in pre-committer topology that ends with a non-forward connection, or there is rescale of committer operator after restarting). Thus the subtaskId should be updated, and they could be passed from the `CheckpointManager`, which could also avoid replicate information between different level of state class~

fapaul replied:
Good catch. I removed the committable summary from the `SubtaskCommittableManager` and only pass the necessary information now.

gaoyunhaii commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/SubtaskCommittableManager.java:
```diff
@@ -0,0 +1,208 @@
+    /**
+     * Returns whether the received number of committables matches the expected number.
+     *
+     * @return boolean
```
The message might need update or remove, also similar to the following methods~

gaoyunhaii commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java:
```diff
@@ -0,0 +1,243 @@
+ * @param <CommT> type of committable
+ */
+@Internal
+public class CommittableCollector<CommT> {
```
Is it possible we also change this class to `CommittableManager`? Since `Collector` seems to already have another meaning (something like output) in Flink project and also it would be consistency with the other two levels.

fapaul replied:
We already have a `CommittableManager` that denotes committables during batch execution. Maybe `CommittableCoordinator` is a better name although coordinator is also taken inside of Flink. I am afraid most general terms already have meaning in the code-base somewhere. :)

gaoyunhaii replied:
Thanks for the explanation! Let's keep the current names for now~

fapaul commented:
@flinkbot run azure

gaoyunhaii commented:
Very thanks @fapaul for the update and explanation!

For the `execute` method, sorry I has some misunderstandings initially that the clear happens before compile happens, but it should indeed be happen after the generate() is called. However, there seems still cases that the transformation is not cleared, like users call `StreamExecutionEnvironment#getExecutionPlan()` ? 



For the Batch exchange, I think perhaps with branches (namely pass the runtime mode to SinkExpander and avoid create the failover region for streaming mode), we could avoid creating the batch exchange in advance for streaming mode, thus we won't need to rely on the behavior of the blocking edge under streaming mode~?

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java:
```diff
@@ -0,0 +1,151 @@
+    /** Mapping of subtask id to {@link SubtaskCommittableManager}. */
+    private final Map<Integer, SubtaskCommittableManager<CommT>> subtasksCommittableManagers;
+
+    private final CommittableCollector<CommT> collector;
```
I think I can just pass the subtaskId and numberOfTasks to the `CheckpointCommittableManager` and remove this way the reverse dependency.

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperator.java:
```diff
@@ -17,95 +17,173 @@
+    }
+
+    private void commitAndEmitCheckpoints() throws IOException, InterruptedException {
+        for (CheckpointCommittableManager<CommT> manager :
```
I tried your suggestion unfortunately it feels like it complicates the solution. First I also need to pass the committer to the commit method
```
List<CommittableMessage> commit(int subtaskId, int totalSubtasks, long lastestCompletedCheckpointId, Committer<CommT> committer);
```
but the biggest problem is that I need to create the `CommittableSummary`s for every checkpoint so I cannot simply return a list of `CommittableMessage`s. The `CommittableManager` is the only object that holds all necessary information to build the correct summary.

gaoyunhaii replied:
Thanks @fapaul for the explanation! I think perhaps `CommittableSummary` is also the subclass of the the `CommittableMessage`  and it should be able to be mixed in the returned list? And passing `committer` is indeed required, but it seems fine from my side~

Also may I have a double confirmation that It seems with the current implementation we'll emit `CommittableSummary` whenever we commit some committables instead of only when checkpointing? It seems to me it is to deal with the case that for unaligned checkpoint, the barrier might jump over all the summary and committable messages~ It should be also the reason we support receive multiple summary from one subtask for the same checkpoint, do I understand right~?

fapaul replied:
> Thanks @fapaul for the explanation! I think perhaps `CommittableSummary` is also the subclass of the the `CommittableMessage` and it should be able to be mixed in the returned list? And passing `committer` is indeed required, but it seems fine from my side~
> 
Good point. To be honest I like the current attraction that we have one central place to coordinate the additions and removals of the `CommittableMessage`: the `CommittableCollector` (I am open for a different name). In addition, we have two different interfaces to actually interact with the committables one is `CommittableManager` that is relevant to commit committables in batch execution mode, and `CheckpointCommittableManager` is used in streaming execution mode to commit committables.



> Also may I have a double confirmation that It seems with the current implementation we'll emit `CommittableSummary` whenever we commit some committables instead of only when checkpointing? It seems to me it is to deal with the case that for unaligned checkpoint, the barrier might jump over all the summary and committable messages~ It should be also the reason we support receive multiple summary from one subtask for the same checkpoint, do I understand right~?

fapaul commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableCollector.java:
```diff
@@ -0,0 +1,243 @@
+    private final NavigableMap<Long, CheckpointCommittableManagerImpl<CommT>>
+            checkpointCommittables;
+    /** Denotes the subtask id the collector is running. */
+    private final int subtaskId;
```
I'd prefer the current solution because during the lifetime of a `CommittableCollector` the `subtaskId` and `numberOfSubtassks` should never change. If I add them as parameters to the commit method it introduces room for errors.

gaoyunhaii replied:
Hi Fabian~ Thanks for the explanation! My main concern comes from that I think current the `CheckpointCommittableManagerImpl` takes two roles: the first is the manager of committables of one checkpoint in memory and another is the state to be stored in the state. For the first role, having `subtaskId` and `numberOfSubtassks` is natural. But for the second role, the two fields are not needed, currently we set it with arbitrary values (0 and 1 respectively), I think this might cause a bit confusion. Thus it seems to me it might be better if we either extract out the two optional fields, or we do not reuse the same object for these two roles~

fapaul replied:
Good point. I hope I resolved the situation by removing the ctor of `CommittableCollector ` and moved the `subtaskId` and `numberOfSubtasks` to the `CommittableCollectorSerializer` that is always created once for the committer operator. So the arbitrary values are not part of the `CommittableCollector` anymore.

fapaul commented:
> For the `execute` method, sorry I has some misunderstandings initially that the clear happens before compile happens, but it should indeed be happen after the generate() is called. However, there seems still cases that the transformation is not cleared, like users call `StreamExecutionEnvironment#getExecutionPlan()` ?

Afaict `getExecutionPlan` is only used within tests and it feels to me like it is some part of Flink's legacy and is not used anymore. I will double-check that.

> For the Batch exchange, I think perhaps with branches (namely pass the runtime mode to SinkExpander and avoid create the failover region for streaming mode), we could avoid creating the batch exchange in advance for streaming mode, thus we won't need to rely on the behavior of the blocking edge under streaming mode~?

Your suggestion is definitely possible but it does not work for exchanges in the sink topology provided by the user. In this case, I would need to move the current logic from `PartitionTransformationTranslator` to `SinkTransformationTranslator`. Perhaps it is a more general issue and the fix should reside in `PartitionTransformationTranslator` rather than only in the sink translation.
Is there a downside when adding it to the `PartitionTransformationTranslator`?

gaoyunhaii commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CommittableManager.java:
```diff
@@ -0,0 +1,52 @@
+ * @param <CommT> type of the committable
+ */
+@Internal
+public interface CommittableManager<CommT> {
```
Hi @fapaul~ Sorry I might not fully got here: It seems currently we are using the method that treats the endOfInput as the checkpoint with id = Long.MAX_VALUE in the implementation? It seems to me that it is sufficient to handle both the streaming and batch cases~? Do we expected to have another `CommittableManager` implementation in the future~?

fapaul replied:
I agree the implementation handles both similarly but the interfaces separate both concerns. Since the operators are in a separate module I think it is a good idea to only expose the interactions through interfaces and do not expose the implementation. 

Having two interfaces allows us in the future to maybe refactor the manager implementation and have different implementations for batch and streaming.

gaoyunhaii replied:
Thanks for the explanation! I still think it would be better to not introduce the interfaces if it is not used for now, but I do not strongly argue for that for now.

fapaul commented:
@gaoyunhaii Thanks for all your help 🙏  From my perspective the last blocking feature about the parallelism is fixed and in regards to the approaching release of Flink 1.15 do you think we can settle with a solution for the internals or maybe open a ticket to fix it after the feature freeze? 

Currently, this PR blocks the migration of the existing sinks inside of Flink that would be great to get it done in 1.15 as well.

gaoyunhaii commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java:
```diff
@@ -20,249 +20,258 @@
+        private <I> DataStream<I> addFailOverRegion(DataStream<I> input) {
+            return new DataStream<>(
+                    executionEnvironment,
+                    new PartitionTransformation<>(
```
We might need to consider the cases like pre-commit operators starts with an operator with different parallelism. Perhaps we need to add a new identity map here.

fapaul replied:
Good point, since all custom topologies are marked as experimental. I would take this as a follow-up.

gaoyunhaii replied:
It is also acceptable to me~

gaoyunhaii commented:
Hi Fabian @fapaul ~ Very thanks for the update! As a whole I'm also agree with that we could move some issues to separate PRs to unblock the following tasks, especially those issues only related to the internal implementation. 

Currently some main remaining issues from my side are as follows:

**1. About setParallelism:** 

I think the current solution to the setParallelism issue is creative, but since it affects the result of `Transformation#getParallelism`, I'm a bit concern on if we have enumerate all the cases. Currently there are two possible issues from my side:

- The semantics of `DataStream#getParallelism` is changed: previously it would always not return -1. This might be fixed by return the env's default if the transformation's parallelism is -1. 
- `TableOperatorWrapperGenerator#calcParallelismAndResource` might have issues if it deals with both transformations with default parallelism and configured parallelism. Sorry I do not see an easy solution to solve this issue.

I tried to see if we could first workaround this issue, but sorry I still not be able to find solutions.

**2. About the repeat transformation**

For the repeat transformation problem, I'm a bit hesitate that why we not remove the newly added transformations instead of the `SinkTransformation`? For example, it is feasible for users to first call `getExecutionPlan()` and then continue to change the properties of the `SinkTransformation`... 

**3. About the commit protocol**

Currently I'm a bit concern if we should keep there are only one summary message for each (checkpoint, subtask) pair, or we allow multiple messages but they must have the save total count. The problem here is that if we allow update the total count, the following operators would not have a method to know if it has received all the committables for one checkpoint. 

Is it possible for now we first throws exception if the `CommitOperator` received the repeat summary message, but keeps the current implementation that it is still possible to emit multiple summary message for a single checkpoint? This poses stricter requirements for `pre-committer toplogy` and `post-committer topology`, thus it would not cause compatibility problem even we change the behavior afterwards.

Also I have tried to think whether unaligned checkpoint + rescaling might cause problems, but currently I do not find issues.

**4. About the PartitionTransformation (Optional)**

> Your suggestion is definitely possible but it does not work for exchanges in the sink topology provided by the user. In this case, I would need to move the current logic from PartitionTransformationTranslator to SinkTransformationTranslator. Perhaps  it is a more general issue and the fix should reside in PartitionTransformationTranslator rather than only in the sink translation.
>  Is there a downside when adding it to the PartitionTransformationTranslator?

I think changing the partition type on-the-fly might cover the possible issues on the callers. I think `PartitionTransformationTranslator` could check and throw exception for impossible cases, but he should not modify it directly. The callers should ensure to avoid the impossible cases. This issue might be postponed since it only related to the implementation for now.

fapaul commented:
> Hi Fabian @fapaul ~ Very thanks for the update! As a whole I'm also agree with that we could move some issues to separate PRs to unblock the following tasks, especially those issues only related to the internal implementation.
> 
> Currently some main remaining issues from my side are as follows:
> 
> **1. About setParallelism:**
> 
> I think the current solution to the setParallelism issue is creative, but since it affects the result of `Transformation#getParallelism`, I'm a bit concern on if we have enumerate all the cases. Currently there are two possible issues from my side:
> 
> * The semantics of `DataStream#getParallelism` is changed: previously it would always not return -1. This might be fixed by return the env's default if the transformation's parallelism is -1.
> * `TableOperatorWrapperGenerator#calcParallelismAndResource` might have issues if it deals with both transformations with default parallelism and configured parallelism. Sorry I do not see an easy solution to solve this issue.

I am not really familiar with this part of the code base but so far I did not see any failing tests regarding the `TableOperatorWrapperGenerator` can you elaborate a bit more why the current solution may cause problems.


> I tried to see if we could first workaround this issue, but sorry I still not be able to find solutions.
> 
> **2. About the repeat transformation**
> 
> For the repeat transformation problem, I'm a bit hesitate that why we not remove the newly added transformations instead of the `SinkTransformation`? For example, it is feasible for users to first call `getExecutionPlan()` and then continue to change the properties of the `SinkTransformation`...

Good point I take a look at this tomorrow.

> **3. About the commit protocol**
> 
> Currently I'm a bit concern if we should keep there are only one summary message for each (checkpoint, subtask) pair, or we allow multiple messages but they must have the save total count. The problem here is that if we allow update the total count, the following operators would not have a method to know if it has received all the committables for one checkpoint.
> 
> Is it possible for now we first throws exception if the `CommitOperator` received the repeat summary message, but keeps the current implementation that it is still possible to emit multiple summary message for a single checkpoint? This poses stricter requirements for `pre-committer toplogy` and `post-committer topology`, thus it would not cause compatibility problem even we change the behavior afterwards.
> 
> Also I have tried to think whether unaligned checkpoint + rescaling might cause problems, but currently I do not find issues.

Afaict the motivation to update the commit summary is exactly for unaligned checkpointing. In case the checkpoint barrier overtakes the records we only emit an incomplete summary and later need to emit a new summary with the updated amount of seen records. 
In general, I am also open to restricting it in the first version and figuring out the unaligned checkpoint use cases later. WDYT?

> 
> **4. About the PartitionTransformation (Optional)**
> 
> > Your suggestion is definitely possible but it does not work for exchanges in the sink topology provided by the user. In this case, I would need to move the current logic from PartitionTransformationTranslator to SinkTransformationTranslator. Perhaps  it is a more general issue and the fix should reside in PartitionTransformationTranslator rather than only in the sink translation.
> > Is there a downside when adding it to the PartitionTransformationTranslator?
> 
> I think changing the partition type on-the-fly might cover the possible issues on the callers. I think `PartitionTransformationTranslator` could check and throw exception for impossible cases, but he should not modify it directly. The callers should ensure to avoid the impossible cases. This issue might be postponed since it only related to the implementation for now.

I think at some point we have to modify the partition transformation. Users might use a partition transformation in their custom sink topologies and they want it to work for batch and streaming mode. So it would be great to allow blocking exchanges in batch and default in streaming mode. 
Throwing an exception would mean to either support batch or stream. Overall, I agree we can revisit this topic after this PR.

gaoyunhaii commented:
Hi Fabian~ 

> I am not really familiar with this part of the code base but so far I did not see any failing tests regarding the TableOperatorWrapperGenerator can you elaborate a bit more why the current solution may cause problems.

Sorry I'm also not expert in table / sql , but from the current implementation of TableOperatorWrapperGenerator, it tries to merge multiple transformations into one `MultipleInputTransformation`, and the parallelism of this one is set with the maximum of all the merged transformations. Previously the transformation might returns env.defaultParallelism in this step, but now it might return -1, suppose we have multiple parallelisms [200, env.default, env.default] and env.default = 300, previously the result would be 300 while now the result becomes 200.

> Afaict the motivation to update the commit summary is exactly for unaligned checkpointing. In case the checkpoint barrier overtakes the records we only emit an incomplete summary and later need to emit a new summary with the updated amount of seen records.
> In general, I am also open to restricting it in the first version and figuring out the unaligned checkpoint use cases later. WDYT?

I think with unaligned checkpointing we should still be able to keep one summary for each checkpoint: each operator could only emit the summary after checkpoint complete notified & received all the summary from each input subtask? If this is feasible we could force the current implementation of pre-commit topology to follow this pattern, but not provide the same promise (only one summary to the same checkpont) to the post-commit topology, then we would not have compatibility no matter which pattern we want in the future~

gaoyunhaii commented:
For the parallelism issue, is it possible that 
1. Before call pre-committer, first modify the default parallelism of env to -1
2. After the call finished, change the default parallelism back.

Then we could know if the parallelism is set and we do not need to change the current behavior of the Transformations ?

fapaul commented:
@gaoyunhaii thanks again for the fruitful discussion. I addressed your concerns

1. Only handling the parallelism overwrite in the SinkTransformationTranslator and deleted the change to the Transformation
2. Delete the added transformations from the environment after the expansion is finished. It allows to run `getExecutionPlan` and `execute` together and allows further configurations to the sink transformation that still apply to the subtransformations.
3. For now I disabled the summary updates https://issues.apache.org/jira/browse/FLINK-25920
4. I left it untouched for now since it also requires followup that is not really related to the Sink

gaoyunhaii reviewed (approved):
Very thanks @fapaul for the update! LGTM and I only have two small comments~

fapaul commented:
@flinkbot run azure
